test/cpp/interop/interop_server.cc

```diff
@@ -349,6 +358,32 @@ class TestServiceImpl : public TestService::Service {
       return Status(grpc::StatusCode::INTERNAL, "Error writing response.");
     }
   }
+
+ private:
+  void RecordServerMetrics(
+      const xds::data::orca::v3::OrcaLoadReport& request_metrics) {
+    // Do not record when zero since it indicates no test per-call report.
+    if (request_metrics.cpu_utilization() > 0) {
+      server_metric_recorder_->SetCpuUtilization(
+          request_metrics.cpu_utilization());
+    }
+    if (request_metrics.mem_utilization() > 0) {
+      server_metric_recorder_->SetMemoryUtilization(
+          request_metrics.mem_utilization());
+    }
+    if (request_metrics.rps_fractional() > 0) {
+      server_metric_recorder_->SetQps(request_metrics.rps_fractional());
+    }
+    std::map<grpc::string_ref, double> named_utilizations;
+    for (const auto& p : request_metrics.utilization()) {
+      const auto& key = *retained_utilization_names_.insert(p.first).first;
+      named_utilizations.emplace(key, p.second);
+    }
+    server_metric_recorder_->SetAllNamedUtilization(named_utilizations);
+  }
+
+  grpc::experimental::ServerMetricRecorder* server_metric_recorder_;
+  std::set<std::string> retained_utilization_names_;
```

Comment:
Just to be safe, I think this needs to be guarded by a mutex, since there could be multiple RPCs in parallel.